README.md
# CNN Route-O-Matic

This is a library for use with [ExpressJS](http://expressjs.com)-based web servers to provide
an extremely versatile routing system.

The Route-O-Matic uses a single configuration set to allow complete build-time OR run-time
configuration of hostnames and routes supported by the server.  Furthermore, the hosts and
routes can be changed at any time without requiring a restart of the server.

#### The library consists of four primary components:

1. The main Route-O-Matic library itself (lib/routeomatic.js), all that needs to be included
   for use.
2. The Host-Table library (lib/host-table.js), which provides the virtual host functionality.
3. The Route-Table library (lib/route-table.js), which provides the route table parsing and
   resolving functionality.
4. The RouteOMatic-Request, or ROM-Request, library (lib/rom-request.js), which provides a
   new request object for use with Route-O-Matic route handling functions.


#### A brief rundown of some key features:

* Fully run-time configurable and re-configurable.
* Supports any number of virtual hosts with unique or shared route tables.
* Route tables can use regular expressions or much faster prefix tree/Trie-based logic.
* Ability to define a default host to use for unconfigured host names.
* Redirects, rewrites, handled routes are all configured as "routes".
* Built-in easy to use proxy logic.


### What is not yet done...

* Documentation needs to be written.
* A test suite needs to be created.



CHANGELOG.md
# CNN Route-O-Matic ChangeLog

## 2018-06-20, Version 0.5.0, @drenter

### Notable changes
- Initial Github release.
- Added ChangeLog.



lib/utils.js
'use strict';

const
    methods = ['GET', 'HEAD', 'POST', 'PUT', 'CHECKOUT', 'COPY', 'DELETE',
        'LOCK', 'MERGE', 'MKACTIVITY', 'MKCOL', 'MOVE', 'M-SEARCH', 'NOTIFY',
        'OPTIONS', 'PATCH', 'PURGE', 'REPORT', 'SEARCH', 'SUBSCRIBE',
        'TRACE', 'UNLOCK', 'UNSUBSCRIBE'],
    writeMethods = ['POST', 'PUT', 'DELETE', 'LOCK', 'MERGE', 'MKACTIVITY',
        'MKCOL', 'MOVE', 'PATCH', 'PURGE', 'UNLOCK', 'UNSUBSCRIBE'];

var utils;


/**
 * Empty function to use with baseLogger
 *
 * @function
 * @private
 * @param {string} _msg - Message to log
 */
function emptyFunc(_msg) {
}


/**
 * Utils module
 *
 * @module utils
 */
utils = {
    /**
     * Basic logging to use if nothing else is defined.
     *
     * @object
     * @public
     */
    baseLogger: {
        debug: emptyFunc,
        error: emptyFunc,
        fatal: function (msg) { console.log(msg); },
        info: emptyFunc,
        warn: emptyFunc
    },

    /**
     * Hostname verification
     *
     * @function
     * @public
     * @param {string} host - Hostname to verify
     * @returns {boolean} - true if valid, false if not
     */
    isHostnameValid: function (host) {
        return (typeof host === 'string' && host.length !== 0 && host.search(/^[\w\-]+(\.[\w\-]+)*$/) !== -1);
    },

    /**
     * Method verification
     *
     * @function
     * @public
     * @param {string} method - Method to verify
     * @returns {boolean} - true if valid, false if not
     */
    isMethodValid: function (method) {
        return (typeof method === 'string' && method.length !== 0 && methods.indexOf(method) !== -1);
    },

    /**
     * Write method check
     *
     * @function
     * @public
     * @param {string} method - Method to check
     * @returns {boolean} - true if "write" method, false if not
     */
    isWriteMethod: function (method) {
        return (typeof method === 'string' && method.length !== 0 && writeMethods.indexOf(method) !== -1);
    },

    /**
     * Get port value with suitable default from urlObject
     *
     * @function
     * @public
     * @param {object} url - The urlObject to pull port from
     * @returns {number} - The port number.  If not specified, returns default for proto (80 or 443).
     */
    portFromUrlObject: function (url) {
        return url.port ? url.port : (url.proto === 'https:' ? 443 : 80);
    },

    /**
     * Extract port from full hostname (thanks for making us have to do this ExpressJS)
     *
     * @function
     * @public
     * @param {string} host - The host string
     * @param {string} proto - The protocol string (http | https)
     * @returns {number} - The port number.  If not specified, returns default for proto (80 or 443).
     */
    extractPortFromHost: function (host, proto) {
        let port,
            ppos;

        if (host.charAt(0) === '[') {
            // IPv6 address
            ppos = host.indexOf(']') + 1;
            if (ppos === 0 || host.charAt(ppos) !== ':') {
                ppos = -1;
            }
        } else {
            ppos = host.indexOf(':');
        }
        if (ppos !== -1) {
            let pstring = host.substring(ppos);
            port = parseInt(pstring.substring(1), 10);
            if (port <= 0) {
                port = (proto === 'https') ? 443 : 80;
            }
        } else {
            port = (proto === 'https') ? 443 : 80;
        }
        return port;
    },

    /**
     * Check request hostname, port, protocol, and method against relevant route matching values, if set.
     *
     * @function
     * @public
     * @param {object} req - The request object
     * @param {object} r - The route object
     * @returns {boolean} - true if matching or unset, false if not a match
     */
    doRuntimeChecks: function (req, r) {
        return ((r.methodMatch.length !== 0 && r.methodMatch !== req.method) ||
            (r.allowWrite !== true && r.methodMatch.length === 0 && writeMethods.indexOf(req.method) !== -1) ||
            (r.portMatch !== 0 && r.portMatch !== req.port) ||
            (r.hostMatch.length !== 0 && r.hostMatch !== req.hostname) ||
            (r.protoMatch.length !== 0 && r.protoMatch !== req.protocol)) ? false : true;
    },

    /**
     * Substitution Parser
     *
     * @function
     * @public
     * @param {string} orig - Original string to parse
     * @param {object} subs - Substitution object
     * @returns {string} - String with substitutions made
     */
    substitute: function (orig, subs) {
        let ns = orig;

        if (typeof ns === 'string' && typeof subs === 'object' && subs !== null) {
            for (let se, ss = ns.indexOf('%'), sw; ss !== -1;) {
                if ((se = ns.indexOf('%', ss + 1)) === -1) {
                    break;
                }
                sw = ns.slice(ss + 1, se);
                if (sw.length !== 0 && typeof subs[sw] === 'string') {
                    if (++se >= ns.length) {
                        ns = (ss === 0 ? '' : ns.slice(0, ss)) + subs[sw];
                        break;
                    }
                    ns = (ss === 0 ? '' : ns.slice(0, ss)) + subs[sw] + ns.slice(se);
                    ss = ns.indexOf('%', ss + subs[sw].length);
                } else {
                    ss = se;
                }
            }
        }

        return ns;
    },

    /**
     * Shallow copy headers from one response object to another
     *
     * @function
     * @public
     * @param {object} copyTo - The response object to copy headers to
     * @param {object} copyFrom - The headers object to copy from
     */
    cloneResponseHeaders: function (copyTo, copyFrom) {
        for (let key in copyFrom) {
            if (copyFrom.hasOwnProperty(key)) {
                copyTo.setHeader(key, copyFrom[key]);
            }
        }
    },

    /**
     * Merge two header objects
     *
     * @function
     * @public
     * @param {object} base - Base header object
     * @param {object} extra - Added header object
     * @throws {error} - If header values are invalid
     * @returns {object} - Merged header object
     */
    mergeHeaders: function (base, extra) {
        let h,
            mh = {};

        if (typeof base === 'object' && base !== null) {
            for (h in base) {
                if (base.hasOwnProperty(h)) {
                    if (typeof base[h] !== 'string') {
                        throw new Error(`Invalid header "${h}" value!`);
                    }
                    mh[h.toLowerCase()] = base[h];
                }
            }
        }
        if (typeof extra === 'object' && extra !== null) {
            for (h in extra) {
                if (extra.hasOwnProperty(h)) {
                    if (typeof extra[h] !== 'string' && typeof extra[h] !== 'number') {
                        throw new Error(`Invalid header "${h}" value!`);
                    }
                    mh[h.toLowerCase()] = extra[h];
                }
            }
        }
        return mh;
    }
};


module.exports = utils;



lib/host-table.js
'use strict';

const
    utils = require('./utils');


/**
 * HostTable Object constructor
 *
 * @constructor
 * @memberof HostTable
 * @param {object} src - Source data for hosts.
 * @param {object} config - Config object.
 * @param {object} config.env - Environment object
 * @param {object} [config.env.subs] - Environment object containing substitutions
 * @param {object} [config.log] - Shared logger, if used.
 * @param {object} config.routeTables - The route tables object
 */
function HostTable(src, config) {
    let
        routeTables,
        subs = null;

    // Initialize values
    this.count = 0;
    this.hasDefaultHost = false;
    this.hosts = {};

    // Validate host table source object
    if (!Array.isArray(src) || src.length === 0) {
        throw new Error('Invalid or empty host table source data!');
    }

    // Validate config object
    if (typeof config !== 'object' || config === null ||
        typeof config.routeTables !== 'object' || config.routeTables === null) {

        throw new Error('Invalid or empty host table config!');
    }
    this.log = config.log || utils.baseLogger;
    routeTables = config.routeTables;
    subs = (config.env && config.env.subs) || null;

    // Check each host table entry
    for (let h, hc, i = 0, rr; i < src.length; i++) {
        try {
            h = src[i];
            if (typeof h !== 'object' || h === null ||
                !Array.isArray(h.hostnames) || h.hostnames.length === 0 ||
                !Array.isArray(h.routeTables) || h.routeTables.length === 0) {

                throw new Error('Invalid or empty host table entry!');
            }

            hc = {};
            // Get host connection timeout, if set
            hc.timeout = (typeof h.timeout === 'number') ? h.timeout :
                (typeof config.defaults.timeout === 'number' ? config.defaults.timeout : 0);

            // Pull together the headers, proxy headers, and redirect headers for each host
            if ((typeof config.defaults.headers === 'object' && config.defaults.headers !== null) ||
                (typeof h.headers === 'object' && h.headers !== null)) {

                hc.headers = utils.mergeHeaders(config.defaults.headers, h.headers);
            } else {
                hc.headers = null;
            }
            if ((typeof config.defaults.proxyHeaders === 'object' && config.defaults.proxyHeaders !== null) ||
                (typeof h.proxyHeaders === 'object' && h.proxyHeaders !== null)) {

                hc.proxyHeaders = utils.mergeHeaders(config.defaults.proxyHeaders, h.proxyHeaders);
            } else {
                hc.proxyHeaders = null;
            }
            if ((typeof config.defaults.redirectHeaders === 'object' && config.defaults.redirectHeaders !== null) ||
                (typeof h.redirectHeaders === 'object' && h.redirectHeaders !== null)) {

                hc.redirectHeaders = utils.mergeHeaders(config.defaults.redirectHeaders, h.redirectHeaders);
            } else {
                hc.redirectHeaders = null;
            }

            // Check for route tables that match the requested ID
            rr = [];
            for (let j = 0, r; j < h.routeTables.length; j++) {
                r = h.routeTables[j];
                if (typeof r !== 'string' || r.length === 0 ||
                    typeof routeTables[r] !== 'object' || routeTables[r] === null) {

                    throw new Error('Invalid or undefined route table!');
                }
                rr.push(routeTables[r].getResolver());
            }

            // Check the hostname list
            for (let n, k = 0; k < h.hostnames.length; k++) {
                n = h.hostnames[k];

                // Deal with substitution strings...
                if (subs !== null) {
                    n = utils.substitute(n, subs);
                    // If resulting hostname is blank, skip this one...
                    if (n === '') {
                        continue;
                    }
                }

                // Verify hostname is valid
                if (!utils.isHostnameValid(n) && n !== '*') {
                    throw new Error(`Invalid hostname "${n}"!`);
                }

                n = n.toLowerCase();
                if (Array.isArray(this.hosts[n]) && this.hosts[n].length !== 0) {
                    throw new Error(`Multiple host table entries for hostname "${n}"!`);
                }

                this.hosts[n] = {
                    config: hc,
                    routeResolvers: rr
                };

                if (n === '*') {
                    this.hasDefaultHost = true;
                }
                this.count++;
            }
        } catch (err) {
            throw new Error(`Error in host table entry #${i}: ` + (err.message || 'Unknown'));
        }
    }
}


/**
 * Retrieve the host object
 *
 * @memberof HostTable
 * @public
 * @param {string} host - The hostname to use
 * @returns {object} - The host object for the hostname
 */
HostTable.prototype.getHost = function (host) {
    let hn = (typeof host === 'string' && host.length !== 0) ? host.toLowerCase() : '*';

    // If we have a match, return it
    if (typeof this.hosts[hn] === 'object' && this.hosts[hn] !== null) {
        return this.hosts[hn];
    }

    // No matching host, if we have a default, return that.
    if (this.hasDefaultHost === true) {
        return this.hosts['*'];
    }

    // Host not found and no default set
    return null;
};


module.exports = HostTable;



lib/http-error.js
/**
 * Super simple HTTP/HTTPS extended error type
 *
 * Really just a normal JS Error with an additional "statusCode" reflecting
 * the HTTP error code.  Also defaults the message to the default HTTP error
 * message for that code if no message is provided.
 *
 * @module http-error
 */

'use strict';


var errorCodes = {
    100: 'Continue',
    101: 'Switching Protocols',
    102: 'Processing',
    200: 'OK',
    201: 'Created',
    202: 'Accepted',
    203: 'Non-Authoritative Information',
    204: 'No Content',
    205: 'Reset Content',
    206: 'Partial Content',
    207: 'Multi-Status',
    208: 'Already Reported',
    226: 'IM Used',
    300: 'Multiple Choices',
    301: 'Moved Permanently',
    302: 'Moved Temporarily',
    303: 'See Other',
    304: 'Not Modified',
    305: 'Use Proxy',
    306: 'Switch Proxy',
    307: 'Temporary Redirect',
    308: 'Permanent Redirect',
    400: 'Bad Request',
    401: 'Unauthorized',
    402: 'Payment Required',
    403: 'Forbidden',
    404: 'Not Found',
    405: 'Method Not Allowed',
    406: 'Not Acceptable',
    407: 'Proxy Authentication Required',
    408: 'Request Timeout',
    409: 'Conflict',
    410: 'Gone',
    411: 'Length Required',
    412: 'Precondition Failed',
    413: 'Payload Too Large',
    414: 'URI Too Long',
    415: 'Unsupported Media Type',
    416: 'Range Not Satisfiable',
    417: 'Expectation Failed',
    418: 'I\'m a teapot',
    421: 'Misdirected Request',
    422: 'Unprocessable Entity',
    423: 'Locked',
    424: 'Failed Dependency',
    426: 'Upgrade Required',
    428: 'Precondition Required',
    429: 'Too Many Requests',
    431: 'Request Header Fields Too Large',
    451: 'Unavailable For Legal Reasons',
    500: 'Internal Server Error',
    501: 'Not Implemented',
    502: 'Bad Gateway',
    503: 'Service Unavailable',
    504: 'Gateway Timeout',
    505: 'HTTP Version Not Supported',
    506: 'Variant Also Negotiates',
    507: 'Insufficient Storage',
    508: 'Loop Detected',
    510: 'Not Extended',
    511: 'Network Authentication Required'
};


/**
 * Utility function to return default error text for a specific error code
 *
 * @function
 * @private
 * @param {number} code - The error code to use
 * @returns {string} - The error message for the given code
 */
function getErrorMessage(code) {
    return (typeof errorCodes[code] === 'string') ? errorCodes[code] : 'Unknown';
}


/**
 * HttpError Object constructor
 *
 * @constructor
 * @memberof http-error
 * @param {string|number} message - Error message, or error status code (if number).
 * @param {number} [status] - Error status code (if not set with message).
 */
function HttpError(message, status) {
    var lpart = (new Error()).stack.match(/[^\s]+$/);

    if (typeof message === 'number') {
        status = message;
        message = undefined;
    } else if (typeof status === 'number' || Number(status) > 0) {
        status = Number(status);
    } else if (typeof message === 'string' && (status = parseInt(message, 10)) > 0) {
        message = (message.length > 3) ? message : undefined;
    }
    this.statusCode = (status > 0 && status < 600) ? status : 500;
    this.message = (typeof message === 'string') ? message : getErrorMessage(this.statusCode);
    this.stack = `${this.name} at ${lpart}`;
}

Object.setPrototypeOf(HttpError, Error);
HttpError.prototype = Object.create(Error.prototype);
HttpError.prototype.name = 'HttpError';
HttpError.prototype.message = '';
HttpError.prototype.statusCode = 0;
HttpError.prototype.constructor = HttpError;

module.exports = HttpError;



lib/trie-route.js
'use strict';

var doRuntimeChecks = require('./utils').doRuntimeChecks;


/**
 * TrieRoute Object constructor
 *
 * @constructor
 * @memberof TrieRoute
 */
function TrieRoute() {
    this.trie = {};
}


/**
 * Add a path/controller to a trie
 * @memberof TrieRoute
 * @public
 * @param {string} path - The path to add to the trie, matched as a substring unless it ends with '#'
 * @param {varies} data - The object or function to return when this pattern is matched
 * @throws Error on failure
 */
TrieRoute.prototype.add = function (path, data) {
    function inject(word, node, depth) {
        let first = word.charAt(depth);

        // If at end of pattern
        if (first === '#' || word.length === depth) {
            let val = (first === '#') ? '|X' : '|W';  // |W = substring match, |X = full string

            if (!(val in node)) {
                node[val] = {};
            }
            if (typeof node[val].val !== 'undefined') {
                throw new Error(`Duplicate or overlapping route!  Failed to add "${path}" because another route resolves the same path or a substring of it (at char ${depth}).`);
            } else {
                node[val].val = data;
            }
        } else {
            if (!(first in node)) {
                node[first] = {};
            }
            inject(word, node[first], depth + 1);
        }
    }

    if (typeof path === 'undefined' || path === null || path === '') {
        throw new Error('Invalid blank/empty route.  Failed to add.');
    }
    inject(path, this.trie, 0);
};


/**
 * Search the trie to see if a match for the given path is found.
 * @memberof TrieRoute
 * @public
 * @param {string} path - The path to find in the trie
 * @param {object} req - The request object to use to compare hostname/protocol, if necessary
 * @returns {object} - The match object on match or null if no match
 */
TrieRoute.prototype.find = function (path, req) {
    function parseWord(word, node, depth) {
        let idx;

        if ('|W' in node && doRuntimeChecks(req, node['|W'].val) === true) {
            // Substring match, we don't need to look any farther
            return {
                data: node['|W'].val,
                match: word.slice(0, depth)
            };
        }
        idx = word.charAt(depth);
        if (idx === '') {
            // Complete word match or not
            if ('|X' in node && doRuntimeChecks(req, node['|X'].val) === true) {
                return {
                    data: node['|X'].val,
                    match: word.slice(0, depth)
                };
            }
            return null;
        }
        return (idx in node) ? parseWord(word, node[idx], depth + 1) : null;
    }

    return (typeof path === 'string' && path.length !== 0) ? parseWord(path, this.trie, 0) : null;
};


module.exports = TrieRoute;



lib/routeomatic.js
/**
 * Route-o-matic -- A festival of routing, redirecting, and rewriting joy for ExpressJS.
 */

'use strict';

const
    defaultAllowUndefinedHandler = false,
    defaultAllowWrite = false,
    defaultNormalizeUrls = false,
    defaultRedirectCode = 302,
    defaultReduceRedirectCode = 301,
    defaultRetryLimit = 20,
    defaultRemoveDoubleSlashes = false,
    defaultTimeout = 20000,
    HostTable = require('./host-table'),
    RomRequest = require('./rom-request'),
    RouteTable = require('./route-table'),
    utils = require('./utils');


/**
 * RouteOMatic object constructor
 *
 * @constructor
 * @param {object} envConf - Environment configuration object
 * @param {object} [envConf.env] - Environment configuration current settings object, optional
 * @param {object} [envConf.env.conds] - Environment configuration conditionals object, optional
 * @param {object} [envConf.env.subs] - Environment configuration substitutions object, optional
 * @param {object} envConf.routeHandlers - Route handlers namespace object
 * @param {object} [envConf.logger] - Shared logger object, optional
 * @param {object} [envConf.requestLogger] - Shared logger object for use in request logging, optional
 * @param {object} [envConf.onSent] - Post response sent function, optional
 * @param {object} hostConf - Host configuration object
 * @param {object} hostConf.defaults - Default host settings object
 * @param {object} [hostConf.defaults.headers] - Default header values, optional
 * @param {array} hostConf.hosts - Array of host config objects
 * @param {array} hostConf.routeTables - Array of route tables config objects
 * @param {object|function} server - Server object (top-level Express object or function)
 */
function RouteOMatic(envConf, hostConf, server) {
    // Initialize object values
    this.hostTable = {};
    this.log = envConf.logger || utils.baseLogger;
    this.logReq = envConf.requestLogger || this.log;
    this.valid = false;

    // Verify server and name parameters
    if (typeof envConf !== 'object' || envConf === null) {
        throw new Error('Invalid environment config object!');
    }
    if (server === null || (typeof server !== 'object' && typeof server !== 'function')) {
        throw new Error('Invalid server object!' + typeof server);
    }

    // Process the hosts, routes, and config
    this.hostTable = this.setupAllTheThings(envConf, hostConf);

    // We have something, so attach it the server
    this.valid = true;
    server.all('*', this.handleRouting.bind(this));
}


/**
 * Check to see if this RouteOMatic instance is "valid".
 *
 * @memberof RouteOMatic
 * @public
 * @returns {boolean} - true if valid, false if not valid
 */
RouteOMatic.prototype.isValid = function () {
    return this.valid;
};


/**
 * Get the currently active host table
 *
 * @memberof RouteOMatic
 * @public
 * @returns {object} - The host table object
 */
RouteOMatic.prototype.getHostTable = function () {
    return this.hostTable;
};


/**
 * Setup all the things!
 *
 * @memberof RouteOMatic
 * @private
 * @param {object} envConf - Environment configuration object
 * @param {object} hostConf - Host configuration object
 * @returns {object} - New host table object
 * @throws {Error} - Throws error on failure
 */
RouteOMatic.prototype.setupAllTheThings = function (envConf, hostConf) {
    let config,
        hostTable = {},
        reqSettings = {},
        routeTables = {};

    // Process the environment config object
    try {
        if (typeof envConf !== 'object' || envConf === null) {
            throw new Error('Invalid or empty configuration');
        }
        if (typeof envConf.routeHandlers !== 'object' || envConf.routeHandlers === null) {
            throw new Error('Invalid route handler context object');
        }
        config = {
            routeHandlers: envConf.routeHandlers
        };
        if (typeof envConf.env !== 'undefined') {
            if (envConf.env === null || typeof envConf.env !== 'object') {
                throw new Error('Invalid current environment settings object');
            }
            config.env = envConf.env;
        }
        if (typeof envConf.logger !== 'undefined') {
            if (envConf.logger === null || typeof envConf.logger !== 'object') {
                throw new Error('Invalid logger object');
            }
            config.log = envConf.logger;
        }
        if (typeof envConf.requestLogger !== 'undefined') {
            if (envConf.requestLogger === null || typeof envConf.requestLogger !== 'object') {
                throw new Error('Invalid requestLogger object');
            }
            reqSettings.requestLogger = envConf.requestLogger;
        }
        if (typeof envConf.dnsLookup !== 'undefined') {
            if (envConf.dnsLookup === null || typeof envConf.dnsLookup !== 'function') {
                throw new Error('Invalid DNS lookup object');
            }
            config.dnsLookup = envConf.dnsLookup;
        }
        if (typeof envConf.onSent === 'function') {
            reqSettings.onSent = envConf.onSent;
        } else {
            reqSettings.onSent = null;
        }
        reqSettings.ports = {};
        if (typeof envConf.ports === 'object' && envConf.ports !== null) {
            let i,
                k = Object.keys(envConf.ports),
                p,
                ps;

            for (i = 0; i < k.length; i++) {
                if (typeof k[i] === 'string') {
                    p = parseInt(k[i], 10);
                } else if (typeof k[i] === 'number') {
                    p = k[i];
                } else {
                    p = 0;
                }
                if (p > 0 && p < 65536) {
                    ps = p.toString(10);
                    if (envConf.ports[k[i]].origProto !== 'http' && envConf.ports[k[i]].origProto !== 'https') {
                        throw new Error(`Invalid protocol specified for port ${k[i]}`);
                    }
                    reqSettings.ports[ps] = {
                        portNumber: p,
                        origProto: envConf.ports[k[i]].origProto,
                        origProtoVer: envConf.ports[k[i]].origProtoVer || '1.1'
                    };
                    if (typeof envConf.ports[k[i]].origPort !== 'undefined') {
                        if (typeof envConf.ports[k[i]].origPort !== 'number') {
                            throw new Error(`Invalid original port specified for port ${k[i]}`);
                        }
                        reqSettings.ports[ps].origPortNumber = envConf.ports[k[i]].origPort;
                        reqSettings.ports[ps].origPort = envConf.ports[k[i]].origPort.toString(10);
                    } else {
                        reqSettings.ports[ps].origPortNumber = p;
                        reqSettings.ports[ps].origPort = ps;
                    }
                }
            }
        }
    } catch (err) {
        throw new Error(`Error processing Route-O-Matic environment configuration: ${err.message}`);
    }

    // Process the host config object
    try {
        if (typeof hostConf !== 'object' || hostConf === null) {
            throw new Error('Invalid or empty configuration!');
        }
        if (typeof hostConf.defaults !== 'undefined' && (typeof hostConf.defaults !== 'object' || hostConf.defaults === null)) {
            throw new Error('Invalid default host settings object!');
        }
        if (!Array.isArray(hostConf.hosts) || hostConf.hosts.length === 0) {
            throw new Error('Invalid or empty hosts configuration!');
        }
        if (typeof hostConf.routeTables !== 'object' || hostConf.routeTables === null) {
            throw new Error('Invalid or empty route table configuration!');
        }
        config.defaults = hostConf.defaults || {};
        // Set any unset defaults
        if (typeof config.defaults.allowWrite !== 'boolean') {
            config.defaults.allowWrite = defaultAllowWrite;
        }
        if (typeof config.defaults.allowUndefinedHandler !== 'boolean') {
            config.defaults.allowUndefinedHandler = defaultAllowUndefinedHandler;
        }
        if (typeof config.defaults.normalizeUrls !== 'boolean') {
            config.defaults.normalizeUrls = defaultNormalizeUrls;
        }
        if (typeof config.defaults.redirectCode !== 'number') {
            config.defaults.redirectCode = defaultRedirectCode;
        }
        if (typeof config.defaults.reduceRedirectCode !== 'number') {
            config.defaults.reduceRedirectCode = defaultReduceRedirectCode;
        }
        if (typeof config.defaults.removeDoubleSlashed !== 'boolean') {
            config.defaults.removeDoubleSlashes = defaultRemoveDoubleSlashes;
        }
        if (typeof config.defaults.retryLimit !== 'number') {
            config.defaults.retryLimit = defaultRetryLimit;
        }
        if (typeof config.defaults.timeout !== 'number') {
            config.defaults.timeout = defaultTimeout;
        }
    } catch (err) {
        throw new Error(`Error processing Route-O-Matic configuration: ${err.message}`);
    }
    reqSettings.allowUndefinedHandler = config.defaults.allowUndefinedHandler;
    reqSettings.allowWrite = config.defaults.allowWrite;
    reqSettings.dnsLookup = config.dnsLookup || null;
    reqSettings.normalizeUrls = config.defaults.normalizeUrls;
    reqSettings.redirectCode = config.defaults.redirectCode;
    reqSettings.reduceRedirectCode = config.defaults.reduceRedirectCode;
    reqSettings.removeDoubleSlashes = config.defaults.removeDoubleSlashes;
    reqSettings.requestLogger = this.logReq;
    reqSettings.retryLimit = config.defaults.retryLimit;
    reqSettings.timeout = config.defaults.timeout;

    // Process the route table objects
    try {
        for (let rt in hostConf.routeTables) {
            if (hostConf.routeTables.hasOwnProperty(rt)) {
                if (typeof hostConf.routeTables[rt] !== 'object' || hostConf.routeTables[rt] === null) {
                    throw new Error(`Bad route table configuration (${rt})`);
                }
                hostConf.routeTables[rt].id = rt;

                try {
                    routeTables[rt] = new RouteTable(hostConf.routeTables[rt], config);
                } catch (e) {
                    throw new Error(`Failure processing route table "${rt}": ${e.message}`);
                }
            }
        }
    } catch (err) {
        throw new Error(`Error processing Route-O-Matic route tables: ${err.message}`);
    }
    config.routeTables = routeTables;

    // Process the host objects
    try {
        hostTable = new HostTable(hostConf.hosts, config);
    } catch (err) {
        throw new Error(`Error processing Route-O-Matic host table: ${err.message}`);
    }

    // Successfully setup new routes and hosts
    this.reqSettings = reqSettings;
    this.reqSettings.hostTable = hostTable;
    this.log.debug('Successfully updated hosts and route tables.');
    return hostTable;
};


/**
 * Re-configure with new hosts/routes/settings
 *
 * @memberof RouteOMatic
 * @public
 * @param {object} newEnvConf - New environment configuration object
 * @param {object} newHostConf - New host configuration object
 * @returns {boolean} - true if successful, false if not
 */
RouteOMatic.prototype.reconfigure = function (newEnvConf, newHostConf) {
    try {
        let newHostTable;

        if (typeof newEnvConf !== 'object' || newEnvConf === null) {
            throw new Error('Invalid environment config object!');
        }

        // Try to re-configure with the new settings
        newHostTable = this.setupAllTheThings(newEnvConf, newHostConf);

        if (newHostTable) {
            delete this.hostTable;
            this.hostTable = newHostTable;
            this.valid = true;
            return true;
        }
    } catch (err) {
        this.log.error(`Failed to reconfigure the routes: ${err.message}`);
    }

    return false;
};


/**
 * Handle ExpressJS routing with the Route-O-Matic
 *
 * @memberof RouteOMatic
 * @public
 * @param {object} req - Request object (Express)
 * @param {object} res - Response object (HTTP/HTTPS)
 * @param {function} [next] - Continuation function
 */
RouteOMatic.prototype.handleRouting = function (req, res, next) {
    let romReq = new RomRequest(this.reqSettings);

    romReq.process(req, res, next);
};


module.exports = RouteOMatic;

